prow/cmd/sidecar/main.go

```diff
@@ -0,0 +1,138 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"flag"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/sirupsen/logrus"
+	"k8s.io/test-infra/prow/logrusutil"
+	"k8s.io/test-infra/prow/pod-utils/wrapper"
+
+	"k8s.io/test-infra/prow/pod-utils/gcs"
+)
+
+type options struct {
+	wrapperOptions *wrapper.Options
+	gcsOptions     *gcs.Options
+}
+
+func (o *options) Validate() error {
+	if err := o.wrapperOptions.Validate(); err != nil {
+		return err
+	}
+
+	return o.gcsOptions.Validate()
+}
+
+func gatherOptions() options {
+	o := options{}
+	fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+	o.wrapperOptions = wrapper.BindOptions(fs)
+	o.gcsOptions = gcs.BindOptions(fs)
+	fs.Parse(os.Args[1:])
+	o.gcsOptions.Complete(fs.Args())
+	return o
+}
+
+func main() {
+	o := gatherOptions()
+	if err := o.Validate(); err != nil {
+		logrus.Fatalf("Invalid options: %v", err)
+	}
+
+	logrus.SetFormatter(
+		logrusutil.NewDefaultFieldsFormatter(nil, logrus.Fields{"component": "sidecar"}),
+	)
+
+	// Only start watching file events if the file doesn't exist
+	// If the file exists, it means the main process already completed.
+	if _, err := os.Stat(o.wrapperOptions.MarkerFile); os.IsNotExist(err) {
+		watcher, err := fsnotify.NewWatcher()
+		if err != nil {
+			logrus.WithError(err).Fatal("Could not begin fsnotify watch")
+		}
+		defer watcher.Close()
+
+		group := sync.WaitGroup{}
+
+		group.Add(1)
+		go func() {
+			defer group.Done()
+			for {
+				select {
```

Comment:
we should probably also select on some low frequency ticker here to check for the file

Reply:
this is done now, xref https://github.com/kubernetes/test-infra/pull/7058/files#r171614657 🙃